File: erlang/seminars/shoutcast/src/tree.erl
-module(tree).

-compile(export_all).

new() ->
    seed.


add(V, seed) ->
    {node, V, seed, seed};
add(V, {node, M, L, R}) when V < M ->
    {node, M, add(V, L), R};
add(V, {node, M, L, R})  ->
    {node, M, L, add(V, R)}.    


sum(seed) ->      
    0;
sum({node, V, Left, Right}) -> 
    V + sum(Left)+ sum(Right).

lookup(_, seed) ->    
    false;
lookup(V, {node, V, _, _}) ->
    true;
lookup(V, {node, M, L, _}) when V < M ->
    lookup(V, L);
lookup(V, {node, _, _, R}) ->
    lookup(V, R).

File: lectures/trees/src/tree.ex
defmodule Tree do

  def list_insert(e, []) do
    [e]
  end
  def list_insert(e, [h|t]) when e < h do
    [e,h|t]
  end
  def list_insert(e, [h|t]) do
    [h|list_insert(e, t)]
  end

  def tree_insert(e, :nil) do {:leaf, e} end
  def tree_insert(e, {:leaf, h}=right) when e < h do
    {:node, e, :nil, right}
  end
  def tree_insert(e, {:leaf, _}=left)  do
    {:node, e, left, :nil}
  end  
  def tree_insert(e, {:node, h, left, right}) when e < h do
    {:node, h, tree_insert(e, left), right}
  end
  def tree_insert(e, {:node, h, left, right}) do
    {:node, h, left, tree_insert(e, right)}
  end  
  

  def bench() do

    ls = [10,20,40,80,160,320,640,1280]

    test = fn (i, f) ->
      seq = sequence(i, 10000)
      elem(:timer.tc(fn () -> f.(seq) end),0)
    end
    IO.write("# benchmark of lists and tree \n")

    bench = fn (i) ->
      list = fn (seq) ->
	List.foldr(seq, [], fn (e, acc) -> list_insert(e, acc) end)
      end
      tree = fn (seq) ->
	List.foldr(seq, :nil, fn (e, acc) -> tree_insert(e, acc) end)
      end      
      tl = test.(i, list) 
      tt = test.(i, tree)     
      IO.write("  #{tl}\t\t#{tt}\n")
    end
    
    Enum.map(ls, bench)
    :ok
  end
  
  def sequence(0, _) do
    []
  end
  def sequence(i, t) do
    [:rand.uniform(t) | sequence(i - 1, t)]
  end
  

end
